File: 1.two-sum.py
#
# @lc app=leetcode id=1 lang=python
#
# [1] Two Sum
#

# @lc code=start
from collections import Counter
class Solution:
    def twoSum(self, nums, target):
        #Brute Force
        # Time complexity: O(n^2)
        # Space complexity: O(1)
        # for i in range(len(nums)):
        #     for j in range(i+1, len(nums)):
        #         if nums[i] + nums[j] == target:
        #             ret
        

        #todo The technique of iterate the right and maintain the left
        # Time complexity: O(n)
        # Space complexity: O(n)

        #When we are traversing through the list, we are also storing the information that is ready to use for the left side
        hashmap = {}
        for i in range(len(nums)):
            if target - nums[i] in hashmap:
                return [hashmap[target - nums[i]], i]
            hashmap[nums[i]] = i
        
        #test for the linux
        return None 
# @lc code=end

File: 1.two-sum.java
/*
 * @lc app=leetcode id=1 lang=java
 *
 * [1] Two Sum
 */
import java.util.HashMap;
// @lc code=start
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // HashMap solution
        // HashMap<Integer, Integer> map = new HashMap<>();
        // for( int i = 0; i < nums.length; i++){
        //     int complement = target - nums[i];
        //     if(map.containsKey(complement))
        //         return new int[] {map.get(complement), i};
        //     map.put(nums[i], i);
        // }
        // return new int[] {0, 0};

        // Brute force solution
        for(int i = 0; i < nums.length -1; i++){
            for(int j = i + 1; j < nums.length; j++){
                if(nums[i] + nums[j] == target){
                    return new int[] {i, j};// the {} is the array initializer shorthand
                }
            }
        }
        return new int[] {};

    }
}
// @lc code=end

